### PR TITLE
Morale System (Port From White Dream)

### PR DESCRIPTION
**Pull request by VMSolidus**
# Description

This Feature has been graciously provided for Einstein Engines to port from the White Dream codebase. 

Mood is a system for tracking a character's current Mental State, which fluctuates throughout the round as a result of various events that can modify it. Each consisting of a single line event that can be trivially inserted into any other system, and a yml configured "Moodlet", which is applied to said character. Moodlets can be temporary or permanent, and can also modify a characters mood in either positive or negative directions. Things like, "Being Hungry", "Being Injured", "Petting a cute animal", "Being Hugged", all create a Moodlet.

Mood can provide buffs or debuffs, primarily to movement speed. In fact Mood's movement speed modifier actually completely replaces the movement speed modifiers from Hunger & Thirst. Instead Hunger & Thirst create a negative moodlet that persists until you eat and drink, which _can_ give you a speed penalty. But you might for instance diminish the negative effects by seeking out other positive sources. Or they might just get worse, who knows what could happen?

# Media

Mood takes the form of a series of Moodlets, which modify your character's internal Mood stat. It's kinda like a healthbar, but for your mental state. Whenever you gain a moodlet, it appears in a popup. White text for standard moodlets, red text for negative moodlets. By clicking on your mood icon, text will show up displaying all of your currently active Moodlets. 

https://github.com/user-attachments/assets/3e9420bb-3a43-4d97-9127-31d704c15287

New traits!
image

Permission from Codeowners:
morale code permission

# TODO

- [x] Refactor the Crit Threshold modification, and Movement Speed Modification to make it more granular. 

# Changelog

:cl: VMSolidus & Skubman
- add: The Mood System has been ported from White Dream. Mood acts as a 3rd healthbar, alongside Health and Stamina, representing your character's current mental state. Having either high or low mood can modify certain physical attributes.
- add: Mood modifies your Critical Threshold. Your critical threshold can be increased or decreased depending on how high or low your character's mood is.
- add: Mood modifies your Movement Speed. Characters move faster when they have an overall high mood, and move slower when they have a lower mood.
- add: Saturnine and Sanguine have been added to the list of Mental traits, both providing innate modifiers to a character's Morale.

### CODE REVIEW

**Comment by VMSolidus** [comment]
ALRIGHT, FINISHED AND READY FOR REVIEW. LET'S GO.

**Comment by VMSolidus** [comment]
All tests passed first try, holy shit.

**Comment by VMSolidus** [comment]
Mood System now not only lives in Shared, but it also now has Contest System entries.

image

File: Resources/Locale/en-US/mood/mood-alerts.ftl [hunk 1]
```diff
@@ -0,0 +1,22 @@
+﻿alerts-mood-insane-name = Insane
+alerts-mood-insane-desc = Darkness and hopelessness smolder in my soul, the world is doomed to absolute evil.
+alerts-mood-horrible-name = Horrible
+alerts-mood-horrible-desc = I struggle with pain and fears, my fate is a series of torments and sufferings.
+alerts-mood-terrible-name = Terrible
+alerts-mood-terrible-desc = My life has dried up like blood from a wound, and there is only darkness and despair all around.
+alerts-mood-bad-name = Bad
+alerts-mood-bad-desc = My strength is leaving me, and every day becomes a difficult ordeal.
+alerts-mood-meh-name = Нехорошо
```
Comment (DangerRevolution) [inline Resources/Locale/en-US/mood/mood-alerts.ftl:9]:
```suggestion
alerts-mood-meh-name = Mediocre
```
Sigma kept the Russian name in......

File: Resources/Locale/en-US/mood/mood-alerts.ftl [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+alerts-mood-meh-desc = The world is full of threats and pain, and my hopes are slowly dying.
```
Comment (DangerRevolution) [inline Resources/Locale/en-US/mood/mood-alerts.ftl:10]:
```suggestion
alerts-mood-meh-desc = The world is full of dangers and pain, and my hopes are slowly dying.
```
Not sure, this sentence doesn't translate well into English

File: Resources/Locale/en-US/mood/mood-alerts.ftl [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+alerts-mood-neutral-name = Neutral
+alerts-mood-neutral-desc = I continue on my way, despite threats and hardships, looking for the slightest light in the darkness.
+alerts-mood-good-name = Good
+alerts-mood-good-desc = In this world of suffering, I find a little relief and hope.
+alerts-mood-great-name = Great
+alerts-mood-great-desc = My strength is restored, and the world seems to be the lesser evil and pain.
+alerts-mood-exceptional-name = Exceptional
+alerts-mood-exceptional-desc = I feel strong and hopeful for better days, despite the threats that lurk around me.
```
Comment (DangerRevolution) [inline Resources/Locale/en-US/mood/mood-alerts.ftl:18]:
```suggestion
alerts-mood-exceptional-desc = Strength and hope fills me, despite the threats that lurk around me.
```

File: Resources/Locale/ru-RU/mood/mood.ftl
Comment (DangerRevolution) [inline Resources/Locale/ru-RU/mood/mood.ftl:1]: remove before merge

File: Resources/Locale/ru-RU/mood/mood-alerts.ftl
Comment (DangerRevolution) [inline Resources/Locale/ru-RU/mood/mood-alerts.ftl:1]: remove before merge

**Review by DangerRevolution — CHANGES_REQUESTED** [review]

File: Resources/Textures/Shaders/White/saturationscale.swsl
Comment (DangerRevolution) [inline Resources/Textures/Shaders/White/saturationscale.swsl:1]: base game folder pls

**Comment by VMSolidus** [comment]
I'm going to temporarily mark this as draft, due to technical issues I was forced to revert back to the last working version of this PR, since migrating it to Shared broke the system. I still desire ContestSystem integrations, which does mean a Shared migration, but for now I can focus on other parts of Refactoring Morale until we get a better solution.

**Review by DangerRevolution — DISMISSED** [review]

**Comment by angelofallars** [comment]
> Petting an animal gives them the petting related buff rather than the humanoid.

Fixed with https://github.com/VMSolidus/Einstein-Engines/pull/3

**Review by DangerRevolution — DISMISSED** [review]

**Review by angelofallars — COMMENTED** [review]
- Adding/removing moodlets should not require raising an event. It should be with `AddMoodlet` and `RemoveMoodlet` methods.
- `generic_postiveEffects.yml` -> `generic_positiveEffects.yml`
- In my local testing, the client did not have the `NetMoodComponent` component.
- The event definitions should probably be placed in `MoodSystem.cs`, the same place where they are raised to be consistent with other systems I've seen.

- The moodlet numbers don't seem that impactful. A character that is both starving and thirsty will have 36 mood, which is not enough to trigger a change in movement speed. A blanket doubling the mood change of all need moodlets and all other moodlets whose change is less than 10 would probably be a step in the right direction. 
  - Your speed only slows down when your mood is Terrible or less (20/100) but I think it should start at 40/100.

**Comment by VMSolidus** [comment]
> * Adding/removing moodlets should not require raising an event. It should be with `AddMoodlet` and `RemoveMoodlet` methods.
> 
>     * `generic_postiveEffects.yml` -> `generic_positiveEffects.yml`
> 
>     * In my local testing, the client did not have the `NetMoodComponent` component.
> 
>     * The event definitions should probably be placed in `MoodSystem.cs`, the same place where they are raised to be consistent with other systems I've seen.
> 
>     * The moodlet numbers don't seem that impactful. A character that is both starving and thirsty will have 36 mood, which is not enough to trigger a change in movement speed. A blanket doubling the mood change of all need moodlets and all other moodlets whose change is less than 10 would probably be a step in the right direction.
>       
>       * Your speed only slows down when your mood is Terrible or less (20/100) but I think it should start at 40/100.

Actually It didn't occur to me that the growth starts at 20 or less, I actually wrote the math assuming it started at <50!!!
Because I made it linearly decrease towards -25% below 50, and geometric curve towards +15% above 50

**Comment by VMSolidus** [comment]
I've changed the thresholds for speed up and speed down to 40 & 60, so that speed modification will occur more readily.

**Comment by VMSolidus** [comment]
image

The fuck? Oh, pointless test.

**Comment by VMSolidus** [comment]
> image
> 
> The fuck? Oh, pointless test.

https://github.com/user-attachments/assets/b62f42ea-4e84-4ee9-a96d-4a7089e9fc91

For my own sanity I went and fact checked the test to make sure it's actually a nothingburger. I was right, there's literally no issue with having Mood modifying movement speed interfering with the slipping system. The test actually did have a pointless Assert.

**Review by angelofallars — COMMENTED** [review]
LGTM. Because two negative moodlets can now trigger a speed change, I feel like this is ready to playtest. Then we can work on refactoring, more moodlets, and more traits.

**Review by DangerRevolution — APPROVED** [review]

File: Content.Server/Mood/MoodSystem.cs [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+        else if (args.OldMobState == MobState.Dead && args.NewMobState != MobState.Dead)
+        {
+            var ev = new MoodRemoveEffectEvent("Dead");
+            RaiseLocalEvent(uid, ev);
+        }
+        RefreshMood(uid, component);
+    }
+
+    // <summary>
+    //   Recalculate the mood level of an entity by summing up all moodlets.
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Mood/MoodSystem.cs:184]:
```suggestion
    //      Recalculate the mood level of an entity by summing up all moodlets.
```

File: Content.Server/Mood/MoodSystem.cs [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+    {
+        if (!_config.GetCVar(CCVars.DoMoodSystem)
+            || !Resolve(uid, ref component)
+            || component.CurrentMoodThreshold == MoodThreshold.Dead && !refresh)
+            return;
+
+        var neutral = component.MoodThresholds[MoodThreshold.Neutral];
+        var ev = new OnSetMoodEvent(uid, amount);
+        RaiseLocalEvent(uid, ref ev);
+        if (ev.Cancelled)
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Mood/MoodSystem.cs:224]:
```suggestion

        if (ev.Cancelled)
```

File: Content.Server/Mood/MoodSystem.cs [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+    {
+        return threshold switch
+        {
+            >= MoodThreshold.Good => 1,
+            <= MoodThreshold.Meh => -1,
+            _ => 0
+        };
+    }
+
+    #region HealthStatusCheck
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Mood/MoodSystem.cs:339]:
```suggestion
```

File: Content.Server/Mood/MoodSystem.cs [hunk 7]
```diff
@@ -0,0 +1,416 @@
...
+            {
+                protoId = threshold.Key;
+                value = threshold.Value;
+            }
+
+        var ev = new MoodEffectEvent(protoId);
+        RaiseLocalEvent(uid, ev);
+    }
+
+    #endregion
```
Comment (DEATHB4DEFEAT) [inline Content.Server/Mood/MoodSystem.cs:360]:
```suggestion
```

File: Content.Shared/Mood/MoodEffectPrototype.cs [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    /// </summary>
+    [DataField]
+    public bool Hidden;
+
+    /// <summary>
+    ///     If mob already has effect of the same category, the new one will replace the old one.
+    /// </summary>
+    [DataField]
+    public string? Category;
+}
```
Comment (DEATHB4DEFEAT) [inline Content.Shared/Mood/MoodEffectPrototype.cs:43]: These should be validated

**Review by DEATHB4DEFEAT — CHANGES_REQUESTED** [review]

**Review by OldDanceJacket — APPROVED** [review]

**Pull request merged**